### DISCUSSION TITLE
Calculated Column

### DISCUSSION THREAD

**Discussion by hugopvl** [directus/directus#5854] [Archive]
guys,

I'm new to directus, and for now I'm loving it, it looks like a Swiss Army knife :)

But I am having some difficulties in creating a calculated column, is it possible? if so what is the most correct way to do it?

I've tried it directly through mysql, I created a trigger that calculates the column total_corte = [quantity * value]. but when I create a new item it gives me an error. 

tell me that the field in the table does not exist, but in reality it does exist ...

![image](https://user-images.githubusercontent.com/81326296/119363211-74cab700-bca5-11eb-951e-03acaa9da7c1.png)

![image](https://user-images.githubusercontent.com/81326296/119363465-bce9d980-bca5-11eb-8b88-10012f9f00c1.png)

**Comment by rijkvanzanten** [comment]
Directus currently doesn't support calculated columns / views yet 👍🏻

**Reply by hugopvl** [reply]
Hi again, and thanks for reply, but its possible to check this information ?

https://github.com/directus/directus/discussions/2787


the shared link for benhaynes is off but he tell is possible do sometinhg with custom hook: ?? 

Thanks.

**Reply by rijkvanzanten** [reply]
This is the correct link https://docs.directus.io/guides/api-hooks/ 👍🏻

**Comment by Bischel** [comment]
You could use hooks API for this. [https://docs.directus.io/guides/api-hooks/](https://docs.directus.io/guides/api-hooks/)
Something like this:

```javascript
module.exports = function registerHook({ exceptions }) {
	return {
		'items.create.before': async function (input, {collection, payload}) {
			if (collection !== 'test_collection') return input;
			payload['count'] = parseInt(payload['number_1']) + parseInt(payload['number_2']);
			console.log(payload);
			return input;
		}
	};
};
```

**Reply by hugopvl** [reply]
hi Bischel, the link is off ! : https://docs.directus.io/guides/api-hooks/

**Reply by neo7-studio-web** [reply]
Updated link : https://docs.directus.io/extensions/hooks/

**Comment by hugopvl** [comment]
Guys, 
only one question,  
I’m new in all this ☺️
 in the manual it says that the default folder for creating hooks is: I presume: directus/extensions/ hooks but I don't have that folder in the extensions directory, do I have to create it?

i am using the latest version of directus v9 rc 68

and by the way, is the name of the hook my choice? then some type of activation is required or it is automatic.

thanks for listening

**Reply by Bischel** [reply]
Yes, just create that directory.
/extensions/hooks/**hook-id**/index.js

**hook-id** could be any name. No activation is required, but don't forget to restart Directus.

**Comment by hugopvl** [comment]
thanks guys, its working like a charme :) 

but now i have a new problem :P 

this work in a moment you create a new file:

```js
module.exports = function registerHook({ exceptions }) {
        return {
                'items.create.before': async function (input, {collection, payload}) {
                        if (collection !== 'planeamento') return input;
                        payload['total_corte'] = parseInt(payload['valor_corte']) * parseInt(payload['quantidade']);
                        console.log(payload);
                        return input;
                }
        };
};
```

**but i need this work to when you update the file.** 

e try: items.update.before but i give a NaN error... 

```js
module.exports = function registerHook({ exceptions }) {
        return {
                'i**tems.update.before**': async function (input, {collection, payload}) {
                        if (collection !== 'planeamento') return input;
                        payload['total_corte'] = parseInt(payload['valor_corte']) * parseInt(payload['quantidade']);
                        console.log(payload);
                        return input;
                }
        };
};
```

![image](https://user-images.githubusercontent.com/81326296/119498562-aacb7200-bd5d-11eb-9c07-d3805234dc95.png)

**Reply by rijkvanzanten** [reply]
Your function `parseInt(payload['valor_corte']) * parseInt(payload['quantidade']);` results in a NaN, so you're saving an illegal value

**Reply by hugopvl** [reply]
ok, but is correct use **items.update.before**  ?

**Reply by rijkvanzanten** [reply]
Yeah! I think the user didn't submit a value for valor_corte or quantidade when you did that calculation 👍🏻

**Reply by Bischel** [reply]
Look at your console. There is a log with payload that sending into db. Maybe you trying to save not an integer value?

**Comment by hugopvl** [comment]
ok, I already understood the problem, if a value greater than 0 is not placed in the valor_corte I will always get NaN, is it possible to convert NaN to 0? because it will not always be necessary to put a value there, I cannot put this mandatory field.

**Reply by Bischel** [reply]
Check "valor_corte" options in admin. It has options "Nullable" and "Default value"

**Reply by hugopvl** [reply]
guys,
thank you in advance for your availability and attention :)
the situation is this, leaving for now the items.update.before that is not working.

I am currently focused on the following:

in the code below, parseInt always returns an integer and I need to work with decimal numbers, I tried without parseInt and or with parsefloat and it works :)


```
module.exports = function registerHook({ exceptions }) {
        return {
                'items.update.before': async function (input, {collection, payload}) {
                        if (collection !== 'planeamento') return input;
                        payload['total_corte'] = parseFloat(payload['valor_corte']) * parseFloat(payload['quantidade']);
                        console.log(payload);
                        return input;
                }
        };
};
```

the problem is when the valor_corte is not filled in, it always returns a NaN, even if it is in the admin options, cut_value to accept nulls.
But I also found that placing 0 by default in the field valor_corte (admin options), gives error NaN, but if I fill in the fields manually, putting 0 in valor_corte (front office) creates the item normally. strange ?

**Reply by Bischel** [reply]
You should check valor_corte and quantidade for not null before calculation in hook code. If any of them is null - you'll get NaN.
```javascript
if ( payload['valor_corte'] != null && payload['quantidade'] != null )
{
   payload['total_corte'] = parseFloat(payload['valor_corte']) * parseFloat(payload['quantidade']);
}
```
And one more thing: check what kind of delimiter do you use in float: comma or dot. Should be dot.

**Reply by rijkvanzanten** [reply]
The update hook only contains the payload of _updates_. If the user didn't save a new value for `valor_corte`, the value will be `undefined`. You might want to load the original existing item first, then use that for defaults for valor corte / quantidade. Next to that, you might want to program the final calculation a little more defensive. Something like this might do (pseudo code, didn't test it):

```js
module.exports = function registerHook({ exceptions, services }) {
	const { ItemsService } = services;

	return {
		'items.update.before': async function (input, { schema, collection, payload, item }) {
			if (collection !== 'planeamento') return input;

			const service = new ItemsService('planeamento', { schema });

			const existingItem = await service.readOne(item[0]);

			const valorCorte = payload.valor_corte ?? existingItem.valor_corte;
			const quantidade = payload.quantidade ?? existingItem.quantidade;

			const totalCorte = parseFloat(valorCorte) * parseFloat(quantidade);

			if (Number.isNaN(totalCorte) === false) {
				payload['total_corte'] = totalCorte
			}

			return input;
		},
	};
};
```

**Comment by hugopvl** [comment]
thanks,

works perfectly


**to insert:**

```js
module.exports = function registerHook({ exceptions, services }) {
        const { ItemsService } = services;

        return {
                'items.create.before': async function (input, { schema, collection, payload, item }) {
                        if (collection !== 'planeamento') return input;

                        const service = new ItemsService('planeamento', { schema });

                        const existingItem = await service.readOne;

                        const valorCorte = payload.valor_corte ?? existingItem.valor_corte;
                        const quantidade = payload.quantidade ?? existingItem.quantidade;

                        const totalCorte = (valorCorte) * (quantidade);
                      
                        if (Number.isNaN(totalCorte) === false) {
                                payload['total_corte'] = totalCorte
                        }
  
                        return input;
                },
        };
};
```

**to update:**

```js
module.exports = function registerHook({ exceptions, services }) {
	const { ItemsService } = services;

	return {
		'items.update.before': async function (input, { schema, collection, payload, item }) {
			if (collection !== 'planeamento') return input;

			const service = new ItemsService('planeamento', { schema });

			const existingItem = await service.readOne(item[0]);

			const valorCorte = payload.valor_corte ?? existingItem.valor_corte;
			const quantidade = payload.quantidade ?? existingItem.quantidade;

			const totalCorte = (valorCorte) * (quantidade);

			if (Number.isNaN(totalCorte) === false) {
				payload['total_corte'] = totalCorte
			}

			return input;
		},
	};
};
```

**Reply by oreilles** [reply]
You don't need custom api hooks or triggers to do that, MySQL supports computed columns. You just need to add it to your table like that and it will just work:
```sql
ALTER TABLE planeamento ADD COLUMN total_corte DECIMAL AS (valor_corte * quantidade)
```
Here's the reference:
https://dev.mysql.com/doc/refman/5.7/en/create-table-generated-columns.html
https://dev.mysql.com/doc/refman/5.7/en/alter-table-generated-columns.html
